### PR TITLE
EFI-182 Trigger reauthentication when OAuth2 login fails

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

File: pEpForiOS/Util/ErrorSubscriber+OAuth2.swift [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+
+            oauthAuthorizer.authorize(authorizer: oauth2Authorizer,
+                                      accountType: accountType,
+                                      viewController: vc)
+        }
+    }
+
+    func handleOAuth2AuthorizationError(error: Error) -> Bool {
+        guard let displayError = DisplayUserError(withError: error) else {
+            return false
```
Comment (martinb89) [inline pEpForiOS/Util/ErrorSubscriber+OAuth2.swift:61]: why could this return nil?

Comment (ghost) [reply pEpForiOS/Util/ErrorSubscriber+OAuth2.swift:61]: > why could this return nil?

```
    /// Creates a user friendly error to present in an alert or such. I case the error type is not
    /// suitable to display to the user (should fail silently), nil is returned.
    ///
    /// - Parameter error: error to create a DisplayUserError for
    /// - Returns:  nil if you should not bother the user with this kind of error,
    ///             user friendly error otherwize.
    init?(withError error: Error)
```

File: pEpForiOS/Util/ErrorSubscriber+OAuth2.swift [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+        if oauthAuthorizers[accountEmail] == nil {
+            let oauthType = oauthType(scopes: scopes)
+            let vc = UIApplication.currentlyVisibleViewController()
+            let oauthAuthorizer = OAuthAuthorizer()
+            oauthAuthorizers[accountEmail] = oauthAuthorizer
+            oauthAuthorizer.delegate = self
+            let oauth2Authorizer = OAuth2ProviderFactory().oauth2Provider().createOAuth2Authorizer()
+            let accountType = oauthType.accountType()
+
+            oauthAuthorizer.authorize(authorizer: oauth2Authorizer,
```
Comment (martinb89) [inline pEpForiOS/Util/ErrorSubscriber+OAuth2.swift]: should we detect if the OAuth web view is presented already?

File: pEpForiOS/Util/ErrorSubscriber+OAuth2.swift [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+        func handleReauthorization(accountEmail: String, scope: String?) {
+            let scopes = (scope ?? "").components(separatedBy: " ")
+            if scopes.isEmpty {
+                Log.shared.errorAndCrash(message: "A valid OAuth2 scope should be defined for XOAuth2 errors")
+                // Without a scope, we'll have to guess the OAuth2 provider, which is bad,
+                // but it's O365 anyways, right?
+            }
+
+            func handleReAuth() {
+                self.handleReauthentication(accountEmail: accountEmail, scopes: scopes)
```
Comment (martinb89) [inline pEpForiOS/Util/ErrorSubscriber+OAuth2.swift]: Should we guarantee that this runs on the main thread?

File: SubProjects/MessageModel/MessageModel/MessageModel/Interface/OAuth2/OAuth2TokenUpdate.swift [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+                keysAlreadyUpdated.insert(key)
+            }
+        }
+    }
+
+    /// Updates all tokens in credentials for the account represented by the given email with the given access token.
+    static public func updateTokens(accountEmail: String, accessToken: OAuth2AccessTokenProtocol) {
+        let session = Session()
+        session.perform {
+            let moc = session.moc
```
Comment (martinb89) [inline SubProjects/MessageModel/MessageModel/MessageModel/Interface/OAuth2/OAuth2TokenUpdate.swift:31]: If possible, use more guard let to facilitate readability.

File: pEpForiOS/Util/ErrorSubscriber+OAuth2.swift (hunk 2 above)
Comment (ghost) [reply pEpForiOS/Util/ErrorSubscriber+OAuth2.swift]: Will implement.

File: pEpForiOS/Util/ErrorSubscriber+OAuth2.swift (hunk 3 above)
Comment (ghost) [reply pEpForiOS/Util/ErrorSubscriber+OAuth2.swift]: Probably a good idea.

File: SubProjects/MessageModel/MessageModel/MessageModel/Interface/OAuth2/OAuth2TokenUpdate.swift (hunk 4 above)
Comment (ghost) [reply SubProjects/MessageModel/MessageModel/MessageModel/Interface/OAuth2/OAuth2TokenUpdate.swift:31]: Will look at it.

Comment (ghost) [reply SubProjects/MessageModel/MessageModel/MessageModel/Interface/OAuth2/OAuth2TokenUpdate.swift:31]: With guards it looks even uglier. Have combined if-lets, though.

**Review by martinb89 — APPROVED** [review]

**Pull request merged**